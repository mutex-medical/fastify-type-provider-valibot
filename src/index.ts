/* eslint-disable @typescript-eslint/no-explicit-any */
import type { ToJSONSchemaOptions } from '@gcornut/valibot-json-schema';
import { toJSONSchema } from '@gcornut/valibot-json-schema';
import type Ajv from 'ajv';
import type { FastifySchema, FastifySchemaCompiler, FastifyTypeProvider } from 'fastify';
import type { FastifySerializerCompiler, FastifyValidationResult } from 'fastify/types/schema';
import { safeParse, parse } from 'valibot';
import type { BaseSchema, InferInput, AnySchema } from 'valibot';

type FreeformRecord = Record<string, any>;

const defaultSkipList = [
  '/documentation/',
  '/documentation/initOAuth',
  '/documentation/json',
  '/documentation/uiConfig',
  '/documentation/yaml',
  '/documentation/*',
  '/documentation/static/*',
];

export interface ValibotTypeProvider extends FastifyTypeProvider {};

const defaultToJSONSchemaOptions: ToJSONSchemaOptions = {
  undefinedStrategy: 'any',
};

interface Schema extends FastifySchema {
  hide?: boolean;
}

export const createJsonSchemaTransform = ({
  skipList,
  ...opts
}: {
  skipList: readonly string[];
  toJSONSchemaOptions?: ToJSONSchemaOptions;
}) => {
  return ({ schema, url }: { schema: Schema; url: string }) => {
    if (!schema) {
      return {
        schema,
        url,
      };
    }

    const toJSONSchemaOptions = opts.toJSONSchemaOptions || defaultToJSONSchemaOptions;

    const { response, headers, querystring, body, params, hide, ...rest } = schema;

    const transformed: FreeformRecord = {};

    if (skipList.includes(url) || hide) {
      transformed.hide = true;
      return { schema: transformed, url };
    }

    const valibotSchemas: FreeformRecord = { headers, querystring, body, params };

    for (const prop in valibotSchemas) {
      const valibotSchema = valibotSchemas[prop];

      if (valibotSchema && valibotSchema.type && !valibotSchema.properties) {
        transformed[prop] = toJSONSchema({ schema: valibotSchema, ...toJSONSchemaOptions });
      }
    }

    if (response) {
      transformed.response = {};

      for (const prop in response as any) {
        const schema = resolveSchema((response as any)[prop]);

        //@ts-ignore
        if (schema && schema.type && !schema.properties) {
          //@ts-ignore
          const transformedResponse = toJSONSchema({ schema: schema, ...toJSONSchemaOptions });
          transformed.response[prop] = transformedResponse;
        }
      }
    }

    for (const prop in rest) {
      const meta = rest[prop as keyof typeof rest];

      if (meta) {
        transformed[prop] = meta;
      }
    }

    return { schema: transformed, url };
  };
};

export const jsonSchemaTransform = createJsonSchemaTransform({
  skipList: defaultSkipList,
});

export const validatorCompiler = (
  ajvInstance?: Ajv,
  fallbackFunction?: (schema: unknown, data: unknown) => FastifyValidationResult,
) =>
  (({ schema }) => {
    //@ts-ignore
    if (schema.type && !schema.properties) {
      return (data) => {
        try {
          parse(schema, data);
          return { value: data };
        } catch (e) {
          return { error: e };
        }
      };
    }

    if (fallbackFunction) {
      return (data) => {
        try {
          return fallbackFunction(schema, data);
        } catch (e) {
          return { error: e };
        }
      };
    }

    if (ajvInstance) {
      return ajvInstance.compile(schema);
    }

    return (data) => {
      return { value: data };
    };
  }) as FastifySchemaCompiler<AnySchema>;

function hasOwnProperty<T, K extends PropertyKey>(obj: T, prop: K): obj is T & Record<K, any> {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

//@ts-ignore
function resolveSchema(maybeSchema): AnySchema {
  if (hasOwnProperty(maybeSchema, 'type')) {
    return maybeSchema;
  }
  throw new Error(`Invalid schema passed: ${JSON.stringify(maybeSchema)}`);
}

export class ResponseValidationError extends Error {
  public details: FreeformRecord;

  constructor(validationResult: FreeformRecord) {
    super("Response doesn't match the schema");
    this.name = 'ResponseValidationError';
    this.details = validationResult.error;
  }
}
export const serializerCompiler: FastifySerializerCompiler<unknown> = ({
  schema: maybeSchema,
}) => {
  const schema = resolveSchema(maybeSchema);

  if ((schema as AnySchema).type) {
    return (data) => {
      const result = safeParse(schema, data);

      if (result.success) {
        return JSON.stringify(result.output);
      }

      throw new ResponseValidationError(result);
    };
  }

  return (data) => JSON.stringify(data);
};
