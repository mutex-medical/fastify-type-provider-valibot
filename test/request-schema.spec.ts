import type { FastifyInstance, FastifyRequest } from 'fastify';
import Fastify from 'fastify';
import * as v from 'valibot';

import type { ValibotTypeProvider } from '../src';
import { serializerCompiler, validatorCompiler } from '../src';

describe('response schema', () => {
  let app: FastifyInstance;
  beforeAll(async () => {
    const REQUEST_SCHEMA = v.object({
      name: v.string(),
    });

    app = Fastify();
    app.setValidatorCompiler(validatorCompiler());
    app.setSerializerCompiler(serializerCompiler);

    app.after(() => {
      app
        .withTypeProvider<ValibotTypeProvider>()
        .route({
          method: 'GET',
          url: '/',
          schema: {
            querystring: REQUEST_SCHEMA,
          },
          handler: (req: FastifyRequest<{ Querystring: v.InferInput<typeof REQUEST_SCHEMA> }>, res) => {
            res.send({
              name: req.query.name,
            });
          },
        })
        .route({
          method: 'GET',
          url: '/no-schema',
          schema: undefined,
          handler: (req, res) => {
            res.send({
              status: 'ok',
            });
          },
        });
    });

    await app.ready();
  });
  afterAll(async () => {
    await app.close();
  });

  it('accepts correct request', async () => {
    const response = await app.inject().get('/').query({
      name: 'test',
    });

    expect(response.statusCode).toBe(200);
    expect(response.json()).toEqual({
      name: 'test',
    });
  });

  it('accepts request on route without schema', async () => {
    const response = await app.inject().get('/no-schema');

    expect(response.statusCode).toBe(200);
    expect(response.json()).toEqual({
      status: 'ok',
    });
  });

  it('returns 400 on validation error', async () => {
    const response = await app.inject().get('/');

    expect(response.statusCode).toBe(400);
    expect(response.json()).toMatchSnapshot();
  });
});
