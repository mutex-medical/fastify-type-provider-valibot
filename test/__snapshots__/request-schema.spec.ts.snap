// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`response schema returns 400 on validation error 1`] = `
{
  "code": "FST_ERR_VALIDATION",
  "error": "Bad Request",
  "message": "Invalid key: Expected "name" but received undefined",
  "statusCode": 400,
}
`;
